File: Tiny2D/Java/src/net/sapium/tiny/graphics/Entity.java
package net.sapium.tiny.graphics;

import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.ArrayList;

import javax.imageio.ImageIO;

import net.sapium.tiny.utils.Vec2D;

import org.apache.log4j.Logger;

public class Entity implements Drawable {

    private double x, y;
    private double dx, dy;
    private double angle;
    private int width, height;
    private ArrayList<Animation> animations;
    private int currentAnimation;
    
    /**
     * Constructs an entity with no animation
     * 
     * @param filename path to the sprite
     */
    public Entity(String filename) {
        animations = new ArrayList<Animation>();
        
        BufferedImage image;
        try {
            image = ImageIO.read(Entity.class.getResourceAsStream(filename));
            
            animations.add(new Animation(image, 1, image.getWidth(), image.getHeight()));
            this.width = image.getWidth();
            this.height = image.getHeight();
        } catch (IOException e) {
            Logger.getLogger(Entity.class).error("Failed to load image", e);
        }
        
        x = y = 0;
        dx = dy = 0;
        angle = 0;
    }
    
    /**
     * Construct a new Entity object
     * 
     * @param filename path to the sprite sheet
     * @param numFrames number of frames in each animation row in the sprite sheet
     * @param width width of each frame
     * @param height height of each frame
     */
    public Entity(String filename, int[] numFrames, int width, int height) {
        animations = new ArrayList<Animation>();
        
        BufferedImage image;
        try {
            image = ImageIO.read(Entity.class.getResourceAsStream(filename));
            
            for(int i=0; i<numFrames.length; i++){
                animations.add(new Animation(image.getSubimage(0, i*height, image.getWidth(), height), 
                                             numFrames[i], width, height));
            }
        } catch (IOException e) {
            Logger.getLogger(Entity.class).error("Failed to load image", e);
        }
        
        this.width = width;
        this.height = height;
        
        x = y = 0;
        dx = dy = 0;
    }
    
    @Override
    public void draw(Graphics2D g) {
        g.rotate(angle, this.getX() + this.getWidth()/2, this.getY() + this.getHeight()/2);
        animations.get(currentAnimation).draw(g, (int) x, (int) y);
        g.rotate(-angle, this.getX() + this.getWidth()/2, this.getY() + this.getHeight()/2);
    }

    public void tick() {
        x += dx;
        y += dy;
        
        animations.get(currentAnimation).tick();
    }
    
    public void play(){
        animations.get(currentAnimation).play();
    }
    
    public void pause() {
        animations.get(currentAnimation).pause();
    }
    
    public void restart() {
        animations.get(currentAnimation).restart();
    }
    
    public void setCurrentAnimation(int animation){
        animations.get(currentAnimation).pause();
        animations.get(currentAnimation).restart();
        
        currentAnimation = animation % animations.size();
    }
    
    public double getRotation() {
        return angle;
    }
    
    public void setRotation(double angle) {
        this.angle = angle;
    }

    public double getX() {
        return x;
    }

    public void setX(double x) {
        this.x = x;
    }

    public double getY() {
        return y;
    }

    public void setY(double y) {
        this.y = y;
    }
    
    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public void setPos(Vec2D v){
        this.x = v.x;
        this.y = v.y;
    }

    public void setVelocity(Vec2D v) {
        this.dx = v.x;
        this.dy = v.y;
    }

    public Vec2D getVelocity() {
        return new Vec2D.Double(dx, dy);
    }
}

File: Tiny2D/C++/lib/graphics/Entity.cpp
#include "graphics/Entity.h"

using namespace Tiny2D;

Entity::Entity(SDL_Renderer *renderer, const char *filename) {
    load(renderer, filename);
}

Entity::Entity(SDL_Renderer *renderer, const char *filename, int numFrames[],
        size_t numAnimations, int width, int height) {
    load(renderer, filename, numFrames, numAnimations, width, height);
}

void Entity::load(SDL_Renderer *renderer, const char *filename, int numFrames[],
        size_t numAnimations, int width, int height) {
    this->load(renderer, filename, 0, width, height);

    this->numFramesArr = numFrames;
    this->numAnimations = numAnimations;

    for(int i = 0; i < numAnimations; i++) {
        this->numFrames += numFramesArr[i];
    }
}


void Entity::draw(SDL_Renderer *renderer) {
    Animation::draw(renderer, this->x, this->y);
}

void Entity::tick(unsigned int dt) {
    this->x += this->dx * dt/1000.0;
    this->y += this->dy * dt/1000.0;

    Animation::tick(dt);

    this->setCurrentFrame(((currentFrame - currentStartFrame) % numFramesArr[currentAnimation]) + currentStartFrame);
}

void Entity::restart() {
    play();
    this->currentFrame = this->currentStartFrame;
}

void Entity::setCurrentAnimation(int animation) {
    this->currentAnimation = animation % numAnimations;

    this->currentStartFrame = 0;
    for(int i = 0; i < currentAnimation; i++) {
        this->currentStartFrame += numFramesArr[i];
    }

    Animation::setCurrentFrame(this->currentStartFrame);
}

void Entity::setCurrentFrame(int frame) {
    Animation::setCurrentFrame(currentStartFrame + (frame % numFramesArr[currentAnimation]));
}

void Entity::nextFrame() {
    this->setCurrentFrame(this->currentFrame - this->currentStartFrame + 1);
}
